### ISSUE TITLE
Drawing offset curves

### ISSUE THREAD

**Issue by bluebear94** [alexheretic/ab-glyph#75]
Are there plans to support drawing [offset curves](https://en.wikipedia.org/wiki/B%C3%A9zier_curve#Offsets_(or_stroking)_of_B%C3%A9zier_curves) for font outlines? I’m currently using `ab_glyph_rasterizer` for rasterizing glyphs from a font and would like to support outlined text as well.

**Comment by alexheretic** [comment]
I don't personally have any plans for such support no.

_ab_glyph_rasterizer_ seems lower level that this concept since i just draws the curves it is given. I guess the missing piece is the logic to generate the offset outline curves from the original glyph outline curves. This could be done outside of _ab_glyph_ & _ab_glyph_rasterizer_.

I'm not 100% sure how you see outlined-text working. Do you want to rasterize both the glyph and then the modified bigger glyph, then render them one on top of the other? If so that does seems possible already.

**Comment by bluebear94** [comment]
I was asking about this since the crate seems to rasterize glyphs by decomposing the outline into line segments, which happen to be easy to offset compared to quadratic and cubic Bézier curves.

**Comment by alexheretic** [comment]
Yes that is an interesting point and reason to bake it into _ab-glyph-rasterizer_. Might be worth a PR.

I guess the only downside would be complicating the rasterizer, but allowing configurable offset without too much extra complexity seems nice to have.

**Comment by bluebear94** [comment]
Offsetting the outlines *would* require storing the list of line segments so that segments that were originally connected stay connected to each other. I think a separate struct that retains such a list instead of drawing them to a buffer might be the best approach, so that `Rasterizer` and (say) `LineBuffer` implement the shared trait

```rust
pub trait CurveDrawer {
  fn draw_line(&mut self, p0: Point, p1: Point);
  fn draw_quad(&mut self, p0: Point, p1: Point, p2: Point) { ... }
  fn draw_cubic(&mut self, p0: Point, p1: Point, p2: Point, p3: Point) { ... }
}
```

Then those who just want to rasterize the original outline could use `Rasterizer` directly, while those who want to rasterize an offset curve could construct a `LineBuffer`, call an `offset` method, and pass it to a `Rasterizer`. Even then, we’d have to worry about the offset curve intersecting itself. I wonder if I should use a completely different approach to this problem.

**Comment by alexheretic** [comment]
It's definitely worth some experimentation. My main worry is if the complexity will be worth it, compared to external logic calculating the offset curves and just keeping _ab_glyph_rasterizer_ as something that can draw the curves.
